src/server/readpool/config.rs

```diff
@@ -48,73 +46,90 @@ pub struct Config {
     pub stack_size: ReadableSize,
 }
 
-impl Default for Config {
-    fn default() -> Config {
-        Config {
-            high_concurrency: DEFAULT_CONCURRENCY,
-            normal_concurrency: DEFAULT_CONCURRENCY,
-            low_concurrency: DEFAULT_CONCURRENCY,
-            max_tasks_high: DEFAULT_MAX_TASKS_PER_CORE * DEFAULT_CONCURRENCY,
-            max_tasks_normal: DEFAULT_MAX_TASKS_PER_CORE * DEFAULT_CONCURRENCY,
-            max_tasks_low: DEFAULT_MAX_TASKS_PER_CORE * DEFAULT_CONCURRENCY,
-            stack_size: ReadableSize::mb(DEFAULT_STACK_SIZE_MB),
-        }
-    }
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Default)]
+#[serde(default)]
+#[serde(rename_all = "kebab-case")]
+/// A high-level struct especially for configuration serialization & deserialization.
+pub struct Config {
+    pub high_concurrency: Option<usize>,
```

Comment:
why Option here?

Reply:
So that we can identify whether or not the configuration item exists. Note that this is the key: we don't provide default values at this level, instead we only check whether or not items exist. After that, we override default values for not-exist items.

Reply:
We have different strategies when doing overrides. For storage, we override missing concurrency to 4, while for coprocessor we override to 8 or CPU_NUM * 0.8. However serde is not clever enough to support different `default()`, so we take it over. Serde only takes care about converting missing values to `None` and we decide what it will become finally.

Reply:
Also note that, the structure at this level "Serde only takes care about converting missing values to None" is different to the structure at this level "we decide what it will become finally". They are different types (however one can be converted to another), which is what I mentioned in the PR body (i.e. Type-DD). Together with the strong type system provided by rust, this ensures that we will never pass None values to the ReadPool.